### PR TITLE
[Enhancement/Feature Issue #47] Added support for KeySharedPolicy for the consumer when in KeyShared mode.

### PR DESCRIPTION
**Pull request by hyperevo**
### Motivation
The pulsar python client lacks support for defining KeyShared behaviour like out of order message delivery and sticky-hash, auto-hash for consumers in KeyShared mode. This PR adds full support. The user can now provide a KeySharedPolicy when starting a consumer with client.subscribe() #47 

The ConsumerConfiguration::KeySharedPolicy and related setter/getter are now exposed to the Python client in this PR.

### Modifications

- Added pybind11 enum for KeySharedMode in src/enums.cc.
- Added pybind11 class for KeySharedPolicy in src/config.cc.
- Modified pybind11 class for ConsumerConfiguration and added function to set KeySharedPolicy and function to read KeySharedPolicy.
- Added KeySharedPolicy wrapper to pulsar/__init__.py. This wrapper handles KeySharedPolicy initialization and does some value validation.
- Added the key_shared_policy parameter to client.subscribe(), some validation, and adding to the config in pulsar/__init__.py.
- Added 4 new tests to test the new KeySharedPolicy functionality to tests/pulsar_test.py

In order to add this functionality I also had to modify the C++ client slightly. The setter function for setting the sticky ranges in the C++ client takes a parameter of the type std::initializer_list<StickyRange> as seen here https://github.com/apache/pulsar-client-cpp/blob/86d66bdb09bdb596a2a5c25b0a09f3d0182d683e/include/pulsar/KeySharedPolicy.h#L99. But std::initializer_list types are not supported and probably won't ever be supported by pybind11. See https://github.com/pybind/pybind11/issues/1302#issuecomment-369090893. So I added an overloaded setter function that takes a parameter of the type  std::vector<StickyRange>. This modification enables us to have python bindings for KeySharedPolicy. I have submitted a PR to the C++ client here https://github.com/apache/pulsar-client-cpp/pull/242. 

This PR won't work until that referenced C++ client PR is merged. 

### Verifying this change
Added 4 new unit tests. Ran them to ensure they pass. 

### Documentation
Added documentation in the form of comments and docstrings to the new classes and parameters that were added.

### CODE REVIEW

**Comment by hyperevo** [comment]
@BewareMyPower Would you be able to help review this PR along with the dependent PR on the C++ client? I see that you are a contributor to both repos.

**Comment by BewareMyPower** [comment]
Sure.

**Comment by hyperevo** [comment]
Thanks!

File: src/config.cc [hunk 1]
```diff
@@ -118,9 +120,21 @@ static ClientConfiguration& ClientConfiguration_setFileLogger(ClientConfiguratio
...
     using namespace py;
 
+    class_<KeySharedPolicy, std::shared_ptr<KeySharedPolicy>>(m, "KeySharedPolicy")
+        .def(init<>())
+        .def("setKeySharedMode", &KeySharedPolicy::setKeySharedMode, return_value_policy::reference)
+        .def("getKeySharedMode", &KeySharedPolicy::getKeySharedMode)
+        .def("setAllowOutOfOrderDelivery", &KeySharedPolicy::setAllowOutOfOrderDelivery, return_value_policy::reference)
+        .def("isAllowOutOfOrderDelivery", &KeySharedPolicy::isAllowOutOfOrderDelivery)
+        .def("setStickyRanges", overload_cast_<StickyRanges>()(&KeySharedPolicy::setStickyRanges), return_value_policy::reference)
+        .def("getStickyRanges", &KeySharedPolicy::getStickyRanges);
```
Comment (BewareMyPower) [inline src/config.cc:136]: We usually use snake case names for methods in Python.

File: src/config.cc [hunk 2]
```diff
@@ -118,9 +120,21 @@ static ClientConfiguration& ClientConfiguration_setFileLogger(ClientConfiguratio
     return conf;
 }
 
+template <typename... Args>
+using overload_cast_ = pybind11::detail::overload_cast_impl<Args...>;
```
Comment (BewareMyPower) [inline src/config.cc:124]: It's better not to use the class under the `detail` namespace, you can perform the type cast manually for overload methods.

File: src/config.cc (hunk 1 above)
Comment (hyperevo) [reply src/config.cc]: Changed them to snake case now.

File: src/config.cc (hunk 2 above)
Comment (hyperevo) [reply src/config.cc]: Done.

File: CMakeLists.txt [hunk 3]
```diff
@@ -148,4 +148,4 @@ add_custom_target(check-format ${BUILD_SUPPORT_DIR}/run_clang_format.py
         ${CLANG_FORMAT_BIN}
         1
         ${BUILD_SUPPORT_DIR}/clang_format_exclusions.txt
-        ${CMAKE_SOURCE_DIR}/src)
+        ${CMAKE_SOURCE_DIR}/src)
```
Comment (BewareMyPower) [inline CMakeLists.txt:151]: Please revert this unrelated change. BTW, it's better to have a line break at the end of a file.

Comment (hyperevo) [reply CMakeLists.txt]: I removed that from the commit now.

**Comment by hyperevo** [comment]
Now that the pulsar cpp client included the code I added in the newest 3.2.0 release, the fixes in this PR will be compatable with it.

**Comment by BewareMyPower** [comment]
The `Oauth2Test.test_base64` seems very flaky, I will rerun it for now.

File: pulsar/__init__.py [hunk 4]
```diff
@@ -1447,6 +1456,73 @@ def policy(self):
...
+        """
+        Returns the actual sticky ranges
+        """
+        return self._policy.get_sticky_ranges()
+
+    def policy(self):
+        """
+        Returns the actual KeySharedPolicy.
+        """
+        return self._policy
```
Comment (BewareMyPower) [inline pulsar/__init__.py:1524]: I think we don't need to expose `_pulsar.KeySharedPolicy` to users.

File: tests/pulsar_test.py [hunk 5]
```diff
@@ -1437,6 +1439,135 @@ def send_callback(res, msg):
...
+                             key_shared_policy=consumer_key_shared_policy)
+
+    def test_keyshared_autosplit(self):
+        client = Client(self.serviceUrl)
+        topic = "my-python-topic-keyshare-autosplit-" + str(time.time())
+        consumer_key_shared_policy = pulsar.ConsumerKeySharedPolicy(
+            key_shared_mode=pulsar.KeySharedMode.AutoSplit,
+            allow_out_of_order_delivery=True,
+        )
+
```
Comment (BewareMyPower) [inline tests/pulsar_test.py:1493]:
```suggestion
```

Remove the line that is consist of all blank characters.

File: tests/pulsar_test.py [hunk 6]
```diff
@@ -1437,6 +1439,135 @@ def send_callback(res, msg):
...
+        
+        consumer = client.subscribe(topic, "my-sub", consumer_type=ConsumerType.KeyShared, consumer_name = 'con-1',
+                                    start_message_id_inclusive=True, key_shared_policy=consumer_key_shared_policy)
+        consumer2 = client.subscribe(topic, "my-sub", consumer_type=ConsumerType.KeyShared, consumer_name = 'con-2',
+                                    start_message_id_inclusive=True, key_shared_policy=consumer_key_shared_policy)
+        producer = client.create_producer(topic)
+
+        for i in range(10):
+            if i > 0:
+                time.sleep(0.02)
```
Comment (BewareMyPower) [inline tests/pulsar_test.py:1501]: Why do we need to sleep 2 milliseconds before each send?

File: tests/pulsar_test.py [hunk 7]
```diff
@@ -1437,6 +1439,135 @@ def send_callback(res, msg):
...
+            allow_out_of_order_delivery=True,
+            sticky_ranges=[(30001, 65535)],
+        )
+        consumer2 = client.subscribe(topic, "my-sub", consumer_type=ConsumerType.KeyShared, consumer_name='con-2',
+                                     start_message_id_inclusive=True, key_shared_policy=consumer2_key_shared_policy)
+        producer = client.create_producer(topic)
+
+        for i in range(10):
+            if i > 0:
+                time.sleep(0.02)
```
Comment (BewareMyPower) [inline tests/pulsar_test.py:1547]: The similar question as I asked before

File: tests/pulsar_test.py [hunk 8]
```diff
@@ -1437,6 +1439,134 @@ def send_callback(res, msg):
...
+            producer.send(b"hello-%d" % i)
+
+        msgs = []
+        while True:
+            try:
+                msg = consumer.receive(100)
+            except pulsar.Timeout:
+                break
+            msgs.append(msgs)
```
Comment (BewareMyPower) [inline tests/pulsar_test.py:1510]:
```suggestion
            msgs.append(msg)
```

File: tests/pulsar_test.py [hunk 9]
```diff
@@ -1437,6 +1439,134 @@ def send_callback(res, msg):
...
```
Comment (BewareMyPower) [inline tests/pulsar_test.py:1556]:
```suggestion
            msgs.append(msg)
```

**Review by BewareMyPower — COMMENTED** [review]
These tests are wrong.

BTW, I think these tests are meaningless that maybe we can remove them. Because they only test two consumers can consume all the messages from a topic when `key_shared_mode`, `allow_out_of_order_delivery`, `sticky_ranges` are configured.

**Comment by BewareMyPower** [comment]
@hyperevo I've left my comments here. Since we're going to release 3.2.0 soon (https://lists.apache.org/thread/4ryrobppjl93gcml63hcnpnjkrn6lrtn), please address the comments ASAP if you want this PR to be included in 3.2.0.

File: tests/pulsar_test.py (hunk 6 above)
Comment (hyperevo) [reply tests/pulsar_test.py:1501]: I added the sleep to keep in line with the other tests. You can see them here https://github.com/apache/pulsar-client-python/blob/39ac8f80d82ca10df5c719b3520985352f578ad4/tests/pulsar_test.py#L861 and here https://github.com/apache/pulsar-client-python/blob/39ac8f80d82ca10df5c719b3520985352f578ad4/tests/pulsar_test.py#LL931C5-L931C5. I am not sure if they are necessary or not. But since they were there in the other tests I assumed there must have been a reason.

Comment (hyperevo) [reply tests/pulsar_test.py:1501]: I will leave the sleeps in for now assuming there is a reason for them in the other tests. But if that is not the case then I can remove them.

**Comment by hyperevo** [comment]
All the comments should be resolved now.

File: tests/pulsar_test.py (hunk 6 above)
Comment (BewareMyPower) [reply tests/pulsar_test.py:1501]: Oh I see. Most of the other `sleep` calls should be removed IMO. In this PR, it's better to remove them. I think we should remove all unnecessary `sleep` calls in other tests in a separated PR.

File: pulsar/__init__.py (hunk 4 above)
Comment (BewareMyPower) [reply pulsar/__init__.py:1524]: Resolve it since the batch receive policy also exposes it.

**Review by BewareMyPower — APPROVED** [review]
Approve it. We can open another PR to remove the sleep calls
